### PR TITLE
Finish check.

### PR DESCRIPTION
**Pull request by LesleyPark**
Finished!

### CODE REVIEW

**Review by dtrypt — COMMENTED** [review]
Is this PR ready to be reviewed?

**Comment by LesleyPark** [comment]
@trustinruss, more or less. Still tinkering with some things because one of the tests in the game_controller spec broke, but if you (or anyone else) has any idea what went wrong there, I'm all ears.

**Comment by dtrypt** [comment]
Is this the error?
```
GamesController games#show action should return a 404 error if the game is not found
Failure/Error: flash[:notice] = 'Black King Check!' if @game.check?('black')
```
Should fix if you add a `return` to the line `render text: 'Not Found', status: :not_found if @game.blank?` so that a game that doesn't exist ends the method rather than running through @game.check. It can't run methods on games that don't exist. The return should have been there from the start.

**Comment by LesleyPark** [comment]
@trustinruss, that fixed it! Thanks (and there's another mistake I hopefully won't be making again, hah)! I'm getting a strange "trailing whitespace error" on config/routes.rb which makes no sense since I haven't touched the file. It goes away when I delete the spacing at line 6 but I feel like we should probably keep that...

**Comment by dtrypt** [comment]
@LesleyPark haven't tested it but I'm guessing you have an indentation at line 6. Delete the indentation rather than the line.
When this PR is ready to be looked at, could you update the description to FINISHED. Thanks!

File: app/models/piece.rb [hunk 1]
```diff
@@ -14,7 +14,7 @@ def move_to!(params)
       # Step 2: Check to see if the piece is of the oppposite color.
       if color != enemy_piece[:color]
         # Step 3: Remove piece from the board.
-        enemy_piece.update_attributes(horizontal_position: nil, vertical_position: nil)
+        enemy_piece.destroy
```
Comment (dtrypt) [inline app/models/piece.rb]: This makes more sense, but if you do this you'll have to fix code that relies on piece position values going to nil throughout the codebase.... and any tests. If this seems like a lot of work then I would just change this line back... but then you would have to add a condition(s) to models/game.rb that excludes nil value pieces from running through the loop in the check?(color) method.

For example, here's one spot that relies on pieces position becoming nil:
In games_controller.rb:
`  def show
    @game = Game.find_by_id(params[:id])

    data = params[:p_type]

    until data.nil?
      @game.pieces.where(type: 'Pawn', vertical_position: 1).first.try(:update_attributes, type: data)
      @game.pieces.where(type: 'Pawn', vertical_position: 8).first.try(:update_attributes, type: data)
      data = params[:p_type] = nil
      redirect_to game_path(@game)
    end `

Comment (koglesby) [reply app/models/piece.rb]: I don't see a problem with this. The example above doesn't rely on pieces position becoming nil; it is referring to the data which comes from a pawn promotion piece selection. It also functions by updating the piece type, as opposed to removing it from the board and replacing it. I didn't notice any test fails from switching to destroy method either..

File: app/controllers/games_controller.rb [hunk 2]
```diff
@@ -13,7 +13,12 @@ def create
 
   def show
     @game = Game.find_by_id(params[:id])
-    render text: 'Not Found', status: :not_found if @game.blank?
+    if @game.blank?
+      render text: 'Not Found', status: :not_found
+      return
+    end
+    flash[:notice] = 'Black King Check!' if @game.check?('black')
+    flash[:notice] = 'White King Check!' if @game.check?('white')
```
Comment (koglesby) [inline app/controllers/games_controller.rb]: Right now the flash notice stays up after the first move to get out of check. To fix that, you can set it to nil when not in check.

 if @game.check?('black')
     flash[:notice] = 'Black King Check!'
    elsif @game.check?('white')
      flash[:notice] = 'White King Check'
    else
     flash[:notice] = nil
    end

File: app/models/piece.rb (hunk 1 above)
Comment (dtrypt) [reply app/models/piece.rb]: Okay. That makes sense. Hopefully we can use the destroy method and have everything still working then.

File: app/views/games/show.html.erb [hunk 3]
```diff
@@ -23,7 +23,7 @@
           data: piece,
           dataType: 'json'
         }).done(function(){
-          window.location.reload();
+          // window.location.reload();
```
Comment (koglesby) [inline app/views/games/show.html.erb]: Don't forget to re-enable the page reload!

**Review by koglesby — CHANGES_REQUESTED** [review]
Nice, triggers the flash notice as it supposed to. Noticed a little bug where the flash message persists on the first move out of check

File: app/models/piece.rb (hunk 1 above)
Comment (LesleyPark) [reply app/models/piece.rb]: @trustinruss, aside from the game_controller spec, the tests were running fine for me as well. And nope, no indentation on config/routes.rb. When I'm on line 6, a left error takes me up to line 5 and a right arrow takes me down to line 7. Deleting the space altogether makes the error go away as does commenting it out, but if there are no objections, I'm just going to leave it as is and ignore Rubocop on that one...

Comment (dtrypt) [reply app/models/piece.rb]: Really? That's strange...I pulled the code in and tested it right now. I'm seeing an indented line 6. If I place the cursor at line 6, it doesn't line up with the cursor at line 9 (which is another empty line). Deleting the indentation fixed the trailing whitespace error for me. If it still doesn't work, I'll fix it in my next pull request.

File: app/controllers/games_controller.rb [hunk 4]
```diff
@@ -13,7 +13,12 @@ def create
 
   def show
     @game = Game.find_by_id(params[:id])
-    render text: 'Not Found', status: :not_found if @game.blank?
```
Comment (dtrypt) [inline app/controllers/games_controller.rb]: Just add return to the front of this line.
`return render text: 'Not Found', status: :not_found if @game.blank?`

File: app/controllers/games_controller.rb (hunk 2 above)
Comment (dtrypt) [reply app/controllers/games_controller.rb]: Nice catch

**Review by dtrypt — CHANGES_REQUESTED** [review]
I think it would be best to rebase to the latest master branch before making the changes we suggested.

File: app/controllers/games_controller.rb (hunk 2 above)
Comment (LesleyPark) [reply app/controllers/games_controller.rb]: @koglesby, interestingly when I did this, Rubocop got hissy about my conditional variables. But when I changed it to:

flash[:notice] = 'Black King Check!' if @game.check?('black')
flash[:notice] = 'White King Check' if @game.check?('white')

sans flash[:notice] = nil, the bug stopped.

Comment (LesleyPark) [inline app/controllers/games_controller.rb]: @koglesby, never mind. Was able to reproduce it. Strange. Looks like it only triggers sometime. Hmm. Should I just overrule Rubocop on this one?

**Comment by LesleyPark** [comment]
@trustinruss, I was going to handle all the new stuff with a merge after everything had been okay-ed. Does that not work here or?

**Review by dtrypt — APPROVED** [review]

File: app/controllers/games_controller.rb [hunk 5]
```diff
@@ -13,17 +13,17 @@ def create
...
-
     data = params[:p_type]
-
+    return render text: 'Not Found', status: :not_found if @game.blank?
     until data.nil?
       @game.pieces.where(type: 'Pawn', vertical_position: 1).first.try(:update_attributes, type: data)
       @game.pieces.where(type: 'Pawn', vertical_position: 8).first.try(:update_attributes, type: data)
       data = params[:p_type] = nil
-      redirect_to game_path(@game)
+      return redirect_to game_path(@game)
```
Comment (sungwon) [inline app/controllers/games_controller.rb:22]: the return here isn't necessary (but probably doesn't hurt either)

File: app/helpers/pieces_helper.rb [hunk 6]
```diff
@@ -1,7 +1,8 @@
 module PiecesHelper
   def capture!(new_x, new_y)
     if color != @chess_piece[:color]
-      @chess_piece.update_attributes(horizontal_position: nil, vertical_position: nil)
+      # @chess_piece.update_attributes(horizontal_position: nil, vertical_position: nil)
```
Comment (sungwon) [inline app/helpers/pieces_helper.rb:4]: next time this file is touched, let's take out the comment.

**Review by sungwon — APPROVED** [review]

File: app/controllers/games_controller.rb (hunk 5 above)
Comment (LesleyPark) [reply app/controllers/games_controller.rb:22]: @trustinruss and I paired on this last night. We found that not having the return there somehow broke the test for pawn promotion in the game controller spec. Adding the return made all tests run cleanly yet again.

**Review by koglesby — APPROVED** [review]

**Pull request merged**